server_modules/file_system.js

```diff
@@ -0,0 +1,133 @@
+var fs = require('fs');
+var Q = require('q');
+var mkdirp = require('mkdirp');
+var path = require('path')
+var utils = require('./utils');
+var args = require('./args');
+
+function getDataFilePath(file) {
+    return exports.resolve(path.dirname(process.argv[1]/*TODO change this to user the args module*/), args.data, file);

```

Comment:
1. `args.data` doesn't exist (args calls it `datadir`).

2. Why would the args module have to be used here? The reason for process.argv[1] is to obtain a reference to where the main scripts lives, such that it can stores its data relative to that (in a `data` subdir).

3. The root is already prepended to the path (by `path.dirname(process.argv[1])`), so by using `exports.resolve()` (instead of `path.resolve()`) it now gets added twice. If it is to prevent against "../"-style attacks, this won't work either. (We'd probably need to wrap `file` with an extra call to `path.resolve()` to prevent against this stuff.)